benchmark/single-source/StringSplitting.swift

```diff
@@ -0,0 +1,314 @@
+    case .scalar:
+      var iter = workload.unicodeScalars.makeIterator()
+      _linesByScalars(source: { iter.next() }, sink: sink)
+    case .utf8:
+      var iter = workload.utf8.makeIterator()
+      _linesByBytes(source: { iter.next() }, sink: sink)
+  }
+}
+
+
+// Inline always to try to ignore any closure stuff
+@inline(__always)
+fileprivate func _linesByCharacters(
+  source characterSource: () throws -> Character?,
+  sink lineSink: (String) -> ()
+) rethrows {
+  var buffer = ""
+  func yield() {
+    lineSink(buffer)
+    buffer.removeAll(keepingCapacity: true)
+  }
+  while let c = try characterSource() {
+    if c.isNewline {
+      yield()
+    } else {
+      buffer.append(c)
+    }
+  }
+
+  // Don't emit an empty newline when there is no more content (e.g. end of file)
+  if !buffer.isEmpty {
+    yield()
+  }
+}
+
+// Inline always to try to ignore any closure stuff
+@inline(__always)
+fileprivate func _linesByScalars(
+  source scalarSource: () throws -> Unicode.Scalar?,
+  sink lineSink: (String) -> ()
+) rethrows {
+	guard #available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *) else {
+		fatalError("unavailable")
+	}
+
+  var buffer: Array<UInt8> = []
+  func yield() {
+    lineSink(String(decoding: buffer, as: UTF8.self))
+    buffer.removeAll(keepingCapacity: true)
+  }
+
+  let _CR = Unicode.Scalar(0x0D)!
+  let _LF = Unicode.Scalar(0x0A)!
+
+  while let first = try scalarSource() {
+    switch first.value {
+    case _CR.value:
+      yield()
+      guard let second = try scalarSource() else { return }
+      if second != _LF { buffer.append(contentsOf: second.utf8) }
+    case 0x000A..<0x000D /* LF ..< CR */: yield()
+    case 0x0085 /* NEXT LINE (NEL) */: yield()
+    case 0x2028 /* LINE SEPARATOR */: yield()
+    case 0x2029 /* PARAGRAPH SEPARATOR */: yield()
+    default: buffer.append(contentsOf: first.utf8)
+    }
+  }
+
+  // Don't emit an empty newline when there is no more content (e.g. end of file)
+  if !buffer.isEmpty {
+    yield()
+  }
+}
+
+// Inline always to try to ignore any closure stuff
+@inline(__always)
+fileprivate func _linesByBytes(
+  source byteSource: () throws -> UInt8?,
+  sink lineSink: (String) -> ()
+) rethrows {
+
+  let _CR: UInt8 = 0x0D
+  let _LF: UInt8 = 0x0A
+
+  var buffer: Array<UInt8> = []
+  func yield() {
+    lineSink(String(decoding: buffer, as: UTF8.self))
+    buffer.removeAll(keepingCapacity: true)
+  }
+
+  /*
+    0D 0A: CR-LF
+    0A | 0B | 0C | 0D: LF, VT, FF, CR
+    E2 80 A8:  U+2028 (LINE SEPARATOR)
+    E2 80 A9:  U+2029 (PARAGRAPH SEPARATOR)
+  */
+  while let first = try byteSource() {
+    switch first {
+    case _CR:
+      yield()
+      // Swallow up any subsequent LF
+      guard let next = try byteSource() else { return }
+      if next != _LF { buffer.append(next) }
+    case 0x0A..<0x0D: yield()
+    case 0xE2:
+      // Try to read: 80 [A8 | A9].
+      // If we can't, then we put the byte in the buffer for error correction
```

Comment:
It might be simpler to have a `case 0xA8, 0xA9:` which removes `[0xE2, 0x80]` from the buffer.
And possibly the same for CRLF — have a `case 0x0A:` which removes `[0x0D]` from the buffer.

Similar example: [InputStream.swift @ 49c342e](https://github.com/apple/swift/blob/49c342e9030378600ed2b21969552d2750fb7851/stdlib/public/core/InputStream.swift)

Reply:
I did try that, and it ran around 1015% slower:

```swift
    case 0xA8...0xA9:
      // Check for E2 80 prior
      let end = buffer.endIndex
      guard buffer.count >= 2, buffer[end - 1] == 0x80, buffer[end - 2] == 0xE2 else {
        buffer.append(first)
        continue
      }
      buffer.removeLast(2)
      yield()
```